File: mobile/android/app/src/main/java/vn/hifpt/clickstream/ClickstreamRepository.kt
package vn.hifpt.clickstream

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.util.Locale

class ClickstreamRepository(private val context: Context) {
    fun loadSessions(assetName: String = "test_data_android.csv"): List<ReplaySession> {
        val source = context.assets.open(assetName).bufferedReader().use { it.readText() }
        val events = if (assetName.lowercase(Locale.US).endsWith(".csv")) {
            parseCsvEvents(source)
        } else {
            parseEvents(source)
        }

        return events
            .filter { it.sessionId.isNotBlank() }
            .groupBy { it.sessionId }
            .map { (sessionId, sessionEvents) ->
                ReplaySession(
                    id = sessionId,
                    events = sessionEvents.sortedWith(
                        compareBy<ClickstreamEvent> { it.timestamp }
                            .thenBy { it.sourceIndex }
                    )
                )
            }
            .sortedByDescending { it.events.size }
    }

    /** Parse JSON event arrays from an asset, file picker, or replay payload. */
    fun parseEvents(json: String): List<ClickstreamEvent> {
        return parseClickstreamJson(json)
    }

    /** Parse the production CSV schema used by android_t3_1k.csv and test_data_android.csv. */
    fun parseCsvEvents(csv: String): List<ClickstreamEvent> {
        val rows = parseCsvRows(csv)
        if (rows.isEmpty()) return emptyList()
        val header = rows.first().map { it.trim().removePrefix("\uFEFF") }
        val index = header.withIndex().associate { it.value to it.index }

        fun value(row: List<String>, vararg names: String): String {
            for (name in names) {
                val idx = index[name]
                if (idx != null) {
                    val v = row.getOrNull(idx).orEmpty().trim()
                    if (v.isNotEmpty()) return v
                }
            }
            return ""
        }

        return rows.drop(1).mapIndexedNotNull { sequence, row ->
            if (row.all { it.isBlank() }) return@mapIndexedNotNull null
            val clientTime = value(row, "client_time", "timestamp")
            val timestamp = parseClientTimeToMillis(clientTime)
            val sessionId = value(row, "session_id")
            if (sessionId.isBlank()) return@mapIndexedNotNull null

            val customerIdRaw = value(row, "customer_id")
            val customerId = if (customerIdRaw.isNotBlank() && customerIdRaw.lowercase(Locale.US) !in setOf("null", "none", "nan")) {
                if (customerIdRaw.endsWith(".0") && customerIdRaw.dropLast(2).all { it.isDigit() }) {
                    customerIdRaw.dropLast(2)
                } else customerIdRaw
            } else null

            ClickstreamEvent(
                sourceIndex = sequence,
                eventId = value(row, "_id", "event_id").ifBlank { "row-$sequence" },
                deviceId = value(row, "device_id"),
                customerId = customerId,
                sessionId = sessionId,
                createdAt = value(row, "device_created_at", "created_at").takeIf { it.isNotBlank() } ?: clientTime,
                timestamp = timestamp,
                key = value(row, "key").ifBlank { "unknown" },
                segment = value(row, "platform", "segmentation_segment", "segment").ifBlank { "Android" },
                name = value(row, "segmentation_name", "name").ifBlank { value(row, "key", "unknown") },
                screenId = value(row, "screen_id", "segmentation_screen_id"),
                durationSeconds = value(row, "dur", "duration").toDoubleOrNull()?.toInt() ?: 0,
                visit = value(row, "visit").takeIf { it.isNotBlank() }
            )
        }
    }
}

private fun parseCsvRows(csv: String): List<List<String>> = buildList {
    val row = mutableListOf<String>()
    val field = StringBuilder()
    var quoted = false
    var index = 0

    fun finishField() {
        row += field.toString()
        field.clear()
    }

    fun finishRow() {
        finishField()
        if (row.isNotEmpty()) add(row.toList())
        row.clear()
    }

    while (index < csv.length) {
        when (val char = csv[index]) {
            '"' -> if (quoted && index + 1 < csv.length && csv[index + 1] == '"') {
                field.append('"')
                index++
            } else {
                quoted = !quoted
            }
            ',' -> if (quoted) field.append(char) else finishField()
            '\n' -> if (quoted) field.append(char) else finishRow()
            '\r' -> Unit
            else -> field.append(char)
        }
        index++
    }
    if (field.isNotEmpty() || row.isNotEmpty()) finishRow()
}

fun parseClickstreamJson(json: String): List<ClickstreamEvent> {
    val trimmed = json.trim()
    if (trimmed.isEmpty()) return emptyList()
    val rows = if (trimmed.startsWith("[")) JSONArray(trimmed) else JSONArray().put(JSONObject(trimmed))
    return ArrayList<ClickstreamEvent>(rows.length()).apply {
        for (index in 0 until rows.length()) {
            add(rows.getJSONObject(index).toEvent(index))
        }
    }
}

private fun JSONObject.toEvent(sequence: Int): ClickstreamEvent {
    val segmentation = optJSONObject("segmentation") ?: JSONObject()
    val eventId = optJSONObject("_id")?.optString("\$oid")
        ?.takeIf { it.isNotBlank() }
        ?: optString("_id").takeIf { it.isNotBlank() }
        ?: "row-$sequence"

    val clientTimeValue = opt("client_time") ?: opt("timestamp")
    val clientTimeText = when (clientTimeValue) {
        is String -> clientTimeValue
        is Number -> clientTimeValue.toLong().toString()
        else -> null
    }
    val createdAt = clientTimeText ?: optJSONObject("created_at")?.optString("\$date")
    val eventTimestamp = parseClientTimeToMillis(clientTimeValue)

    val customerId = if (has("customer_id") && !isNull("customer_id")) {
        val raw = opt("customer_id")?.toString()?.trim()
        if (raw.isNullOrBlank() || raw.lowercase(Locale.US) in setOf("null", "none", "nan")) null
        else if (raw.endsWith(".0") && raw.dropLast(2).all { it.isDigit() }) raw.dropLast(2)
        else raw
    } else null

    return ClickstreamEvent(
        sourceIndex = sequence,
        eventId = eventId,
        deviceId = optString("device_id"),
        customerId = customerId,
        sessionId = optString("session_id"),
        createdAt = createdAt,
        timestamp = eventTimestamp,
        key = optString("key", "unknown"),
        segment = firstNonBlank(
            optString("platform"),
            segmentation.optString("segment"),
            optString("segmentation.segment"),
            "Android"
        ),
        name = firstNonBlank(
            optString("segmentation_name"),
            segmentation.optString("name"),
            optString("segmentation.name"),
            optString("key", "unknown")
        ),
        screenId = firstNonBlank(
            optString("screen_id"),
            segmentation.optString("screen_id"),
            optString("segmentation.screen_id")
        ),
        durationSeconds = 0,
        visit = firstNonBlankOrNull(
            segmentation.opt("visit")?.toString(),
            optString("segmentation.visit")
        )
    )
}

private fun parseClientTimeToMillis(value: Any?): Long = when (value) {
    null -> 0L
    is Number -> value.toLong()
    is String -> {
        val s = value.trim()
        s.toLongOrNull() ?: runCatching { Instant.parse(s).toEpochMilli() }.getOrDefault(0L)
    }
    else -> 0L
}

private fun firstNonBlank(vararg values: String): String =
    values.firstOrNull { it.isNotBlank() } ?: ""

private fun firstNonBlankOrNull(vararg values: String?): String? =
    values.firstOrNull { !it.isNullOrBlank() }

File: mobile/ios/Sources/JourneyClustering/Storage/ClickstreamRepository.swift
import Foundation

/// Repository for parsing production clickstream CSV logs and JSON payloads into ClickstreamEvent models.
public final class ClickstreamRepository: @unchecked Sendable {
    public init() {}

    /// Parse CSV or JSON content into replay sessions grouped by sessionId.
    public static func loadSessions(content: String, isCsv: Bool) -> [ReplaySession] {
        let events = isCsv ? parseCsvEvents(content) : parseJsonEvents(content)
        let filtered = events.filter { !$0.sessionId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

        let grouped = Dictionary(grouping: filtered, by: { $0.sessionId })
        return grouped.map { (sessionId, sessionEvents) in
            let sorted = sessionEvents.sorted { a, b in
                if a.timestamp != b.timestamp { return a.timestamp < b.timestamp }
                return a.sourceIndex < b.sourceIndex
            }
            return ReplaySession(id: sessionId, events: sorted)
        }.sorted { $0.events.count > $1.events.count }
    }

    /// Parse production CSV schema used by mobile clickstream dumps.
    public static func parseCsvEvents(_ csv: String) -> [ClickstreamEvent] {
        let rows = parseCsvRows(csv)
        guard !rows.isEmpty else { return [] }

        let header = rows[0].map { $0.trimmingCharacters(in: .whitespacesAndNewlines).replacingOccurrences(of: "\u{FEFF}", with: "") }
        var indexMap: [String: Int] = [:]
        for (idx, name) in header.enumerated() {
            indexMap[name] = idx
        }

        func val(_ row: [String], _ names: [String]) -> String {
            for name in names {
                if let idx = indexMap[name], idx < row.count {
                    let v = row[idx].trimmingCharacters(in: .whitespacesAndNewlines)
                    if !v.isEmpty { return v }
                }
            }
            return ""
        }

        var events: [ClickstreamEvent] = []
        for (sequence, row) in rows.dropFirst().enumerated() {
            if row.allSatisfy({ $0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }) {
                continue
            }
            let clientTime = val(row, ["client_time", "timestamp"])
            let timestamp = parseClientTimeToMillis(clientTime)
            let sessionId = val(row, ["session_id"])
            if sessionId.isEmpty { continue }

            let customerIdRaw = val(row, ["customer_id"])
            let customerId: String?
            if !customerIdRaw.isEmpty && !["null", "none", "nan"].contains(customerIdRaw.lowercased()) {
                if customerIdRaw.hasSuffix(".0") && customerIdRaw.dropLast(2).allSatisfy({ $0.isNumber }) {
                    customerId = String(customerIdRaw.dropLast(2))
                } else {
                    customerId = customerIdRaw
                }
            } else {
                customerId = nil
            }

            let createdAtRaw = val(row, ["device_created_at", "created_at"])
            let createdAt = !createdAtRaw.isEmpty ? createdAtRaw : clientTime

            let segment = val(row, ["platform", "segmentation_segment", "segment"])
            let name = val(row, ["segmentation_name", "name"])
            let key = val(row, ["key"])

            let event = ClickstreamEvent(
                sourceIndex: sequence,
                eventId: val(row, ["_id", "event_id"]).isEmpty ? "row-\(sequence)" : val(row, ["_id", "event_id"]),
                deviceId: val(row, ["device_id"]),
                customerId: customerId,
                sessionId: sessionId,
                createdAt: createdAt.isEmpty ? nil : createdAt,
                timestamp: timestamp,
                key: key.isEmpty ? "unknown" : key,
                segment: segment.isEmpty ? "iOS" : segment,
                name: name.isEmpty ? (key.isEmpty ? "unknown" : key) : name,
                screenId: val(row, ["screen_id", "segmentation_screen_id"]),
                durationSeconds: Int(Double(val(row, ["dur", "duration"])) ?? 0.0),
                visit: val(row, ["visit"]).isEmpty ? nil : val(row, ["visit"])
            )
            events.append(event)
        }
        return events
    }

    /// Parse JSON arrays or single objects from live SDK streams or test files.
    public static func parseJsonEvents(_ json: String) -> [ClickstreamEvent] {
        let trimmed = json.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let data = trimmed.data(using: .utf8) else { return [] }

        guard let jsonObject = try? JSONSerialization.jsonObject(with: data) else { return [] }

        let rows: [[String: Any]]
        if let arr = jsonObject as? [[String: Any]] {
            rows = arr
        } else if let dict = jsonObject as? [String: Any] {
            rows = [dict]
        } else {
            return []
        }

        return rows.enumerated().map { (sequence, dict) in
            dictToEvent(dict: dict, sequence: sequence)
        }
    }

    private static func dictToEvent(dict: [String: Any], sequence: Int) -> ClickstreamEvent {
        let segmentation = (dict["segmentation"] as? [String: Any]) ?? [:]

        let eventId: String
        if let idDict = dict["_id"] as? [String: Any], let oid = idDict["$oid"] as? String, !oid.isEmpty {
            eventId = oid
        } else if let idStr = dict["_id"] as? String, !idStr.isEmpty {
            eventId = idStr
        } else {
            eventId = "row-\(sequence)"
        }

        let clientTimeValue = dict["client_time"] ?? dict["timestamp"]
        let clientTimeText: String?
        if let s = clientTimeValue as? String {
            clientTimeText = s
        } else if let n = clientTimeValue as? NSNumber {
            clientTimeText = "\(n.int64Value)"
        } else {
            clientTimeText = nil
        }

        let createdAtDate = (dict["created_at"] as? [String: Any])?["$date"] as? String
        let createdAt = clientTimeText ?? createdAtDate
        let timestamp = parseClientTimeToMillis(clientTimeValue)

        let customerId: String?
        if let rawCust = dict["customer_id"] {
            let str = "\(rawCust)".trimmingCharacters(in: .whitespacesAndNewlines)
            if str.isEmpty || ["null", "none", "nan"].contains(str.lowercased()) {
                customerId = nil
            } else if str.hasSuffix(".0") && str.dropLast(2).allSatisfy({ $0.isNumber }) {
                customerId = String(str.dropLast(2))
            } else {
                customerId = str
            }
        } else {
            customerId = nil
        }

        let platform = firstNonBlank([
            dict["platform"] as? String,
            segmentation["segment"] as? String,
            segmentation["segmentation.segment"] as? String,
            "iOS"
        ])

        let name = firstNonBlank([
            dict["segmentation_name"] as? String,
            segmentation["name"] as? String,
            segmentation["segmentation.name"] as? String,
            dict["key"] as? String,
            "unknown"
        ])

        let screenId = firstNonBlank([
            dict["screen_id"] as? String,
            segmentation["screen_id"] as? String,
            segmentation["segmentation.screen_id"] as? String
        ])

        let visit = firstNonBlankOrNil([
            (segmentation["visit"] != nil) ? "\(segmentation["visit"]!)" : nil,
            segmentation["segmentation.visit"] as? String
        ])

        return ClickstreamEvent(
            sourceIndex: sequence,
            eventId: eventId,
            deviceId: dict["device_id"] as? String ?? "",
            customerId: customerId,
            sessionId: dict["session_id"] as? String ?? "",
            createdAt: createdAt,
            timestamp: timestamp,
            key: dict["key"] as? String ?? "unknown",
            segment: platform,
            name: name,
            screenId: screenId,
            durationSeconds: 0,
            visit: visit
        )
    }

    private static func parseCsvRows(_ csv: String) -> [[String]] {
        var rows: [[String]] = []
        var row: [String] = []
        var field = ""
        var quoted = false
        let chars = Array(csv)
        var index = 0

        func finishField() {
            row.append(field)
            field = ""
        }

        func finishRow() {
            finishField()
            if !row.isEmpty { rows.append(row) }
            row = []
        }

        while index < chars.count {
            let char = chars[index]
            switch char {
            case "\"":
                if quoted && index + 1 < chars.count && chars[index + 1] == "\"" {
                    field.append("\"")
                    index += 1
                } else {
                    quoted = !quoted
                }
            case ",":
                if quoted {
                    field.append(char)
                } else {
                    finishField()
                }
            case "\n":
                if quoted {
                    field.append(char)
                } else {
                    finishRow()
                }
            case "\r":
                break
            default:
                field.append(char)
            }
            index += 1
        }
        if !field.isEmpty || !row.isEmpty {
            finishRow()
        }
        return rows
    }

    private static let isoWithMillis: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoStandard: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    public static func parseClientTimeToMillis(_ value: Any?) -> Int64 {
        guard let value = value else { return 0 }
        if let num = value as? NSNumber { return num.int64Value }
        if let s = value as? String {
            let trimmed = s.trimmingCharacters(in: .whitespacesAndNewlines)
            if let direct = Int64(trimmed) { return direct }
            if let date = isoWithMillis.date(from: trimmed) ?? isoStandard.date(from: trimmed) {
                return Int64(date.timeIntervalSince1970 * 1000.0)
            }
        }
        return 0
    }

    private static func firstNonBlank(_ values: [String?]) -> String {
        for v in values {
            if let str = v?.trimmingCharacters(in: .whitespacesAndNewlines), !str.isEmpty {
                return str
            }
        }
        return ""
    }

    private static func firstNonBlankOrNil(_ values: [String?]) -> String? {
        for v in values {
            if let str = v?.trimmingCharacters(in: .whitespacesAndNewlines), !str.isEmpty {
                return str
            }
        }
        return nil
    }
}
